### DISCUSSION TITLE
How to match a contact with a phone number that uses a country code using a search string that does not use the country code?

### DISCUSSION THREAD

**Discussion by mianaliasjad** [vestrel00/contacts-android#258] [Q&A] [answered]
I want to find the details of a saved phone number. I have a phone number "03004481849" which is saved in my contacts. When I try to find it with `broadQuery` it returns null.

```
    val contact = Contacts(context)
        .broadQuery()
        .match(BroadQuery.Match.PHONE)
        .wherePartiallyMatches( "03004481849" )
        .limit(1)
        .find().firstOrNull()
```

But when I try to find this contact using `ContactsContract.PhoneLookup.CONTENT_FILTER_URI` i can find it.

```
   val filterUri = Uri.withAppendedPath(
        ContactsContract.PhoneLookup.CONTENT_FILTER_URI,
        Uri.encode("03004481849")
    )
    val projection = arrayOf(
        BaseColumns._ID,
        ContactsContract.PhoneLookup.DISPLAY_NAME,
        ContactsContract.PhoneLookup.NUMBER,
        ContactsContract.PhoneLookup.PHOTO_URI
    )
    val cursor = context.contentResolver.query(filterUri, projection, null, null, null)
```
It returns the cursor with the result and I can read it. Is it a bug in the library because it can't find it?

**Comment by mianaliasjad** [comment]
I just checked your source code you are using `ContactsContract.CommonDataKinds.Phone.CONTENT_FILTER_URI` to find the contact when we specify `Match.PHONE` in `broadQuery ` on the other hand I used `ContactsContract.PhoneLookup.CONTENT_FILTER_URI` maybe that's the reason?

For more details you can look at this [question](https://stackoverflow.com/questions/3079365/android-retrieve-contact-name-from-phone-number)

**Comment by vestrel00** [comment]
@mianaliasjad, I am currently unable to reproduce this issue using the following code you provided;

```kotlin
   val contact = Contacts(context)
        .broadQuery()
        .match(BroadQuery.Match.PHONE)
        .wherePartiallyMatches( "03004481849" )
        .limit(1)
        .find()
        .firstOrNull()
```


I created a contact with the phone number "03004481849" and I am able to get it.

> on the other hand I used `ContactsContract.PhoneLookup.CONTENT_FILTER_URI` maybe that's the reason?

Maybe. I'm not sure until I can reproduce it.

There are a few things I want you to try to help me figure if this is a bug or not.

First, could you make sure that you have the `android.permission.READ_CONTACTS` when executing the broad query. Try using the `permissions` extensions. It would look something like this;

```kotlin
   val contact = Contacts(context)
        .broadQueryWithPermission()
        .match(BroadQuery.Match.PHONE)
        .wherePartiallyMatches( "03004481849" )
        .limit(1)
        .find()
       .firstOrNull()
```

If permissions is not the issue, please provide me more data. It would really help if you give me the logs to the Contacts Provider database tables. Before you copy and paste the logs here in public for everyone to see, please make sure there are no sensitive data there. You can make sure of this by using fake data.

Read this page on how to do this; https://vestrel00.github.io/contacts-android/debug/debug-contacts-provider-tables/

Essentially, you just have to use the `Context.logContactsProviderTables()` in the `debug` module of this project. Filter the Logcat with "ContactsDebug" to see all the logs.

**Reply by vestrel00** [reply]
Also, please tell me what device and API version you are testing on and see if you can reproduce the issue in a google pixel emulator/phone.

**Comment by mianaliasjad** [comment]
Yes, Contacts permission is granted. 

I am testing on the Samsung s22 ultra
SDK_INT = 31

**Reply by vestrel00** [reply]
Okay good. It means there might be a bug! Please give me the debug logs when you can ^

**Comment by mianaliasjad** [comment, marked as answer]
Ok i found a hint for you:

Let's suppose a number with a country code is saved "+923123456789" but when i try to find it without country code "03123456789" the library is unable to find it. 

But my method using `ContactsContract.PhoneLookup.CONTENT_FILTER_URI` i can find it.

**Reply by vestrel00** [reply]
This is a really good hint!!! I'll see if I can reproduce it.

**Reply by vestrel00** [reply]
Okay. So I am definitely able to reproduce it. So I will definitely create an issue for this!

However, I am not sure if the issue should be a bug or a new feature. I verified that the AOSP Contacts app, Google Contacts app, and Samsung Contacts app have the same "issue". The contact with a phone number of "+923123456789" will not be returned if the search string is  "03123456789".

The `BroadQuery` API that this library provides is not using `ContactsContract.PhoneLookup.CONTENT_FILTER_URI` because it's purpose is to emulate the behavior of the other major Contacts app, which it is doing. I need to think of what to do in terms of "solving" this problem. Either;

- we consider it a bug in the `BroadQuery` API and patch it there
- or it will be a new feature that will be added in a new API, which I might call `PhoneLookupQuery`

For now, I will create the issue and include it in the next release; **[0.2.4](https://github.com/vestrel00/contacts-android/discussions/248)**

I will mark this thread as the answer and post the issue link when I have created it. Thanks a lot for raising this discussion!

**Reply by mianaliasjad** [reply]
> 

I think we should introduce it as a "new feature" because as you said the purpose of the BroadQuery is to emulate the behavior of the other major Contacts app.

**Reply by vestrel00** [reply]
I agree! Okay. Here is the issue; #259 

It will be included in the next release; **[0.2.4](https://github.com/vestrel00/contacts-android/discussions/248)**

Documentation and example usage of the new API will be included 😁 

Thanks for all your help @mianaliasjad! This library will be better because of our discussion here!

**Comment by mianaliasjad** [comment]
For future readers, currently, there is no proper way in this library to find the details of contact with the given phone number. Let's suppose you want the details of contact with the number "0347".
```

val contact = Contacts(context)
        .broadQuery()
        .match(BroadQuery.Match.PHONE)
        .wherePartiallyMatches("0347")
        .limit(1)
        .find().firstOrNull()
```

It will return the details of all the numbers that end with "****0347" which is not desired because we want details for "0347" exact number.  So as @vestrel00 described we should add a new feature called `PhoneLookupQuery `.

**Reply by vestrel00** [reply]
This is true. Again, it is most appropriately a missing feature more than a bug. The scenario you described with searching using "0347" and in turn matches "****0347" can be reproduced in the AOSP Contacts app, Google Contacts app, and Samsung Contacts app. 

What is expected behavior for you may not be the expected behavior for others. I bet there would be other people that would create bugs/issues if the `BroadQuery` API did the things we talked about in this discussion.

As agreed upon, we will keep the `BroadQuery` API the way it is now. It will return the same results as the other major contacts apps.

A new, specialized API will be introduced for this type of phone number searches ❤️

**Reply by mianaliasjad** [reply]
> This is true. Again, it is most appropriately a missing feature more than a bug. The scenario you described with searching using "0347" and in turn matches "****0347" can be reproduced in the AOSP Contacts app, Google Contacts app, and Samsung Contacts app.
> 
> What is expected behavior for you may not be the expected behavior for others. I bet there would be other people that would create bugs/issues if the `BroadQuery` API did the things we talked about in this discussion.
> 
> As agreed upon, we will keep the `BroadQuery` API the way it is now. It will return the same results as the other major contacts apps.
> 
> A new, specialized API will be introduced for this type of phone number searches ❤️

You are right `BroadQuery` is doing its job. `BroadQuery` can be used in a search bar functionality of a contacts app. But let's suppose I am creating a dialer app and there is an incoming call with the number "0347" and I want to find the name of the caller to show. in that case, I need the record of "0347" exact number not of numbers that end with "******0347" so yeah `PhoneLookupQuery` is the solution for this.

**Reply by vestrel00** [reply]
Yes. I 100% agree with you. This is a critical missing feature of this library. Thanks for bringing it up!
